packages/common/src/utils/sdkMigrationUtils.ts

```diff
@@ -0,0 +1,162 @@
+import { detailedDiff } from 'deep-object-diff'
+import { isEmpty } from 'lodash'
+
+import { ErrorLevel, ReportToSentryArgs } from '~/models/ErrorReporting'
+import { FeatureFlags, RemoteConfigInstance } from '~/services/remote-config'
+
+export type CheckSDKMigrationArgs<T> = {
+  legacy: T
+  migrated?: T | Error
+}
+
+export class SDKMigrationFailedError extends Error {
+  public endpointName: string
+  public innerMessage: string
+  public legacyValue: unknown
+  public migratedValue: unknown
+
+  constructor({
+    endpointName,
+    innerMessage,
+    legacyValue,
+    migratedValue
+  }: {
+    endpointName: string
+    innerMessage: string
+    legacyValue?: unknown
+    migratedValue?: unknown
+  }) {
+    super(`Diff ${endpointName} failed: ${innerMessage}`)
+    this.name = 'SDKMigrationFailedError'
+    this.endpointName = endpointName
+    this.innerMessage = innerMessage
+    this.legacyValue = legacyValue
+    this.migratedValue = migratedValue
+  }
+}
+
+/** Compares a legacy and migrated response, which must be the same shape. For
+ * literal values, will do a strict equals. For objects, will do a deep diff.
+ * Throws `SDKMigrationFailedError` if there is a difference between the two responses.
+ */
+export const compareSDKResponse = <T extends object>(
+  { legacy, migrated }: CheckSDKMigrationArgs<T>,
+  endpointName: string
+) => {
+  // Migrated is an error, skip the diff
+  if (migrated instanceof Error) {
+    throw new SDKMigrationFailedError({
+      endpointName,
+      innerMessage: 'Migrated response was error',
+      legacyValue: legacy,
+      migratedValue: migrated
+    })
+  }
+  // Both object-like, perform deep diff
+  if (typeof legacy === 'object' && typeof migrated === 'object') {
+    const diffValue = detailedDiff(legacy, migrated)
+    if (
+      !isEmpty(diffValue.added) ||
+      !isEmpty(diffValue.deleted) ||
+      !isEmpty(diffValue.updated)
+    ) {
+      throw new SDKMigrationFailedError({
+        endpointName,
+        innerMessage: 'Legacy and migrated values differ',
+        legacyValue: legacy,
+        migratedValue: migrated
+      })
+    }
+  }
+  // Not object like, perform strict equals
+  if (legacy !== migrated) {
```

Comment:
should this be an `else if`? don't we only want to do strict equals if they're not objects?

Reply:
Oh jeez, I definitely happy-pathed this function. Good catch.